src/Native/Unix/System.Native/pal_random.c

```diff
@@ -2,109 +2,219 @@
-                {
-                    // Another thread has already set the rand_des
-                    close(fd);
-                }
-            }
-            else if (errno == ENOENT)
-            {                
-                sMissingDevURandom = true;
+                continue;
             }
+
+            assert(false && "getrandom() has failed");
+            return false;
         }
 
-        if (rand_des != -1)
+        buffer += n;
+        length -= (size_t)n;
+    }
+    while(length);
+
+    return true;
+}
+
+static void FillWithGetRandom(uint8_t* buffer, int32_t bufferLength)
+{
+    if (!TryFillWithGetRandom(buffer, bufferLength))
+    {
+        FillWithMrand(buffer, bufferLength);
+    }
+}
+#endif // HAVE_SYS_GETRANDOM
+
+static bool TryFillWithDevUrandom(uint8_t *buffer, int32_t bufferLength)
+{
+    int fd = GetUrandomFileDescriptor();
+    size_t length = (size_t)bufferLength;
+
+    do
+    {
+        ssize_t n = read(fd, buffer, length);
+
+        if (n == -1)
         {
-            int32_t offset = 0;
-            do
+            if (errno == EINTR)
             {
-                ssize_t n = read(rand_des, buffer + offset , (size_t)(bufferLength - offset));
-                if (n == -1)
-                {
-                    if (errno == EINTR)
-                    {
-                        continue;
-                    }
-
-                    assert(false && "read from /dev/urandom has failed");
-                    break;
-                }
-
-                offset += n;
+                continue;
             }
-            while (offset != bufferLength);
+
+            assert(false && "read from /dev/urandom has failed");
+            return false;
         }
+
+        buffer += n;
+        length -= (size_t)n;
     }
+    while(length);
 
-    if (!sInitializedMRand)
+    return true;
+}
+
+static void FillWithDevUrandom(uint8_t* buffer, int32_t bufferLength)
+{
+    if (!TryFillWithDevUrandom(buffer, bufferLength))
     {
-        srand48(time(NULL));
-        sInitializedMRand = true;
+        FillWithMrand(buffer, bufferLength);
     }
+}
 
-    // always xor srand48 over the whole buffer to get some randomness
-    // in case /dev/urandom is not really random
+typedef void (*RandomFillFunction)(uint8_t *buffer, int32_t length);
 
-    for (int i = 0; i < bufferLength; i++)
+static RandomFillFunction ResolveRandomFillFunction(void)
+{
+#if HAVE_SYS_GETRANDOM
+    if (IsGetRandomAvailable())
     {
-        if (i % 4 == 0)
-        {
-            num = lrand48();
-        }
+        return FillWithGetRandom;
+    }
+#endif // HAVE_SYS_GETRANDOM
 
-        *(buffer + i) ^= num;
-        num >>= 8;
+    if (GetUrandomFileDescriptor() >= 0)
+    {
+        return FillWithDevUrandom;
     }
-#endif // HAVE_ARC4RANDOM_BUF
+
+    return FillWithMrand;
+}
+
+static RandomFillFunction g_randomFillFunction = FillWithMrand;
+
+__attribute__((constructor(65535)))
```

Comment:
The static constructors are problematic. We try to avoid them:
- Ordering problems when you get more of them
- Performance overhead of doing unnecessary work during startup
- They are not standard C/C++.

Reply:
Resolved.